### ISSUE TITLE
Adam/Adagrad gives different derivatives when initializing globaly and localy

### ISSUE THREAD

**Issue by s4sarath** [tensorflow/tensorflow#29675]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- TensorFlow installed from (source or binary):
- TensorFlow version (2.0):
- Python version:(3.6)

**Describe the current behavior**
Adam and Adagrad, both gives different losses, when initialized once (globally) and initialized every time train function is called. 

````
#### Locally
def train2(model, inputs, outputs, learning_rate):
    trainable_variables = [model.W, model.b]
    with tf.GradientTape() as t:
        current_loss = loss(model(inputs), outputs)
    optimizer = tf.keras.optimizers.Adagrad(
    learning_rate=learning_rate,
    initial_accumulator_value=0.1,
    epsilon=1e-07,
)
    gradients = t.gradient(current_loss,  trainable_variables)    
    optimizer.apply_gradients(zip(gradients, trainable_variables))
#----------------------------------------------------------------------------------------------------
#### Globally

optimizer_global = tf.keras.optimizers.Adagrad(
    learning_rate=0.1,
    initial_accumulator_value=0.1,
    epsilon=1e-07,
)

def train(model, inputs, outputs, learning_rate):
    trainable_variables = [model.W, model.b]
    with tf.GradientTape() as t:
        current_loss = loss(model(inputs), outputs)
    
    gradients = t.gradient(current_loss,  trainable_variables)    
    optimizer_global.apply_gradients(zip(gradients, trainable_variables))
````

**Describe the expected behavior**
We expect same loss from both train and train2. ( Is it because, some internal parameters are changing inside the optimizer, because SGD optimizer works fine in both the cases).

**Code to reproduce the issue**
The plot of derivatives are provided in the image . Full code to reproduce is in the .ipynb file.
![tf](https://user-images.githubusercontent.com/10637096/59318400-9151cc80-8ce4-11e9-9b65-1712a86d2218.png)

````
#!/usr/bin/env python
# coding: utf-8



import tensorflow as tf
tf.compat.v1.set_random_seed(1)
import numpy as np
tf.__version__



class Model(object):
  def __init__(self):
    # Initialize variable to (5.0, 0.0)
    # In practice, these should be initialized to random values.
    self.W = tf.Variable(5.0)
    self.b = tf.Variable(0.0)
    
  def __call__(self, x):
    return self.W * x + self.b
  
model = Model()
def loss(predicted_y, desired_y):
    return tf.reduce_mean(tf.square(predicted_y - desired_y))

TRUE_W = 3.0
TRUE_b = 2.0
NUM_EXAMPLES = 1000

inputs  = tf.random.normal(shape=[NUM_EXAMPLES])
noise   = tf.random.normal(shape=[NUM_EXAMPLES])
outputs = inputs * TRUE_W + TRUE_b + noise # 3 x + 2 + noise()

import matplotlib.pyplot as plt
get_ipython().run_line_magic('matplotlib', 'inline')
plt.scatter(inputs, outputs, c='b')
plt.scatter(inputs, model(inputs), c='r')
plt.show()

def train2(model, inputs, outputs, learning_rate):
    trainable_variables = [model.W, model.b]
    with tf.GradientTape() as t:
        current_loss = loss(model(inputs), outputs)
    optimizer = tf.keras.optimizers.Adam(learning_rate, beta_1=0.9, beta_2=0.98, 
                                         epsilon=1e-9)
    gradients = t.gradient(current_loss,  trainable_variables)    
    optimizer.apply_gradients(zip(gradients, trainable_variables))
    
#-----------------------------------------------------------------------------------------------

optimizer_global = tf.keras.optimizers.Adam(learning_rate=0.1, beta_1=0.9, beta_2=0.98, 
                                         epsilon=1e-9)
def train(model, inputs, outputs, learning_rate):
    trainable_variables = [model.W, model.b]
    with tf.GradientTape() as t:
        current_loss = loss(model(inputs), outputs)
    
    gradients = t.gradient(current_loss,  trainable_variables)    
    optimizer_global.apply_gradients(zip(gradients, trainable_variables))

model = Model()

# Collect the history of W-values and b-values to plot later
Ws, bs = [], []
epochs = range(100)
for epoch in epochs:
  Ws.append(model.W.numpy())
  bs.append(model.b.numpy())
  current_loss = loss(model(inputs), outputs)

  train(model, inputs, outputs, learning_rate=0.1)
  print('Epoch %2d: W=%1.2f b=%1.2f, loss=%2.5f' %
        (epoch, Ws[-1], bs[-1], current_loss))

# Let's plot it all
plt.plot(epochs, Ws, 'r',
         epochs, bs, 'b')
plt.plot([TRUE_W] * len(epochs), 'r--',
         [TRUE_b] * len(epochs), 'b--')
plt.legend(['W', 'b', 'true W', 'true_b'])
plt.show()
  

#---------------------------------------------------------

model = Model()

# Collect the history of W-values and b-values to plot later
Ws, bs = [], []
epochs = range(100)
for epoch in epochs:
  Ws.append(model.W.numpy())
  bs.append(model.b.numpy())
  current_loss = loss(model(inputs), outputs)

  train2(model, inputs, outputs, learning_rate=0.1)
  print('Epoch %2d: W=%1.2f b=%1.2f, loss=%2.5f' %
        (epoch, Ws[-1], bs[-1], current_loss))

# Let's plot it all
plt.plot(epochs, Ws, 'r',
         epochs, bs, 'b')
plt.plot([TRUE_W] * len(epochs), 'r--',
         [TRUE_b] * len(epochs), 'b--')
plt.legend(['W', 'b', 'true W', 'true_b'])
plt.show()
  ````

**Comment by gadagashwini** [comment]
Was able to reproduce the reported issue on Colab with Tensorflow 2.0.0.beta0. Thanks!

**Comment by tanzhenyu** [comment]
Adam is stateful, if you create it everytime calling train2, its behavior is meant to be different than if you create a global optimizer.

**Comment by s4sarath** [comment]
@tanzhenyu - Thanks. Which is the preferred method ? Initializing globally or at each time train function is called.

**Comment by tanzhenyu** [comment]
Initializing globally

**Comment by yfq512** [comment]
@s4sarath  Hi, I have a question about other script ，When run ...'tf.compat.v1.set_random_seed(1)', it will raise error'module 'tensorflow.compat' has no attribute 'v1'', Can you help me? Waiting your reply!

**Comment by tanzhenyu** [comment]
Depending on your tf version. compat.v1 is introduced only after 1.14. So if you're in 1.14 you shouldn't see the error. Otherwise it would pop up.
